### PR TITLE
Reflection-based monoid solver

### PR DESCRIPTION
**Pull request by oisdk**
I have started packaging up the [ring solver I wrote](https://oisdk.github.io/agda-ring-solver/README.html) into smaller components that I'll turn into pull requests over the next few weeks, and this is the first of those.

It's a small reflection-based monoid solver, which uses the same interface as the ring solver. I'd like to get an idea of how to organize this code in relation to the standard library, which I can then base the much larger ring solver on.

As it is, it's not ready for merging (I'm waiting on the `--safe` version of the Reflection module #799).

It contains a few different parts:

* A re implementation of the `Expr` type from Algebra.Solver.Monoid.
* Some generic reflection tools and helpers which should probably go in Reflection once #799 is done.
* The actual solver interface.
* An extra reasoning combinator for setoid reasoning.
* Two examples for the README files on how to use the solver.

### CODE REVIEW

File: src/Algebra/Solver/Monoid/Reflection.agda [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+  ⟦_⇓⟧′ : Expr Carrier → Carrier → Carrier
+  ⟦ x ⊙ y  ⇓⟧′ z = ⟦ x ⇓⟧′ (⟦ y ⇓⟧′ z)
+  ⟦ ⓔ      ⇓⟧′ y = y
+  ⟦ ⟦ x ↑⟧ ⇓⟧′ y = x ∙ y
+
+  ⟦_⇓⟧ : Expr Carrier → Carrier
+  ⟦ x ⇓⟧ = ⟦ x ⇓⟧′ ε
+
+  open import Relation.Binary.Reasoning.Setoid setoid
```
Comment (MatthewDaggitt) [inline src/Algebra/Solver/Monoid/Reflection.agda]: Please could you import this at the top of the file (e.g. as `SetoidReasoning`) and then open it here.

File: src/Algebra/Solver/Monoid/Reflection.agda [hunk 2]
```diff
@@ -0,0 +1,185 @@
+{-# OPTIONS --without-K --safe #-}
```
[the remark this answers is not in the document]
Comment (oisdk) [reply src/Algebra/Solver/Monoid/Reflection.agda]: I was wondering why the CI failed!

File: src/Algebra/Solver/Monoid/Reflection.agda [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+open import Agda.Builtin.Reflection
+
+----------------------------------------------------------------------
+-- The Expr type with homomorphism proofs
+----------------------------------------------------------------------
+
+infixl 7 _⊙_
+data Expr {a} (A : Set a) : Set a where
+  _⊙_ : Expr A → Expr A → Expr A
```
[the remark this answers is not in the document]
Comment (oisdk) [reply src/Algebra/Solver/Monoid/Reflection.agda]: I actually might change the constructors for Expr (in my editor, the characters are messing up the spacing)

File: README/MonoidSolver.agda [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+open import Relation.Binary.Reasoning.Setoid (setoid ℕ)
+
+example₁ : ∀ x y z → (x + y) + z ≡ x + (y + z) + 0
+example₁ x y z = solve +-0-monoid
+
+example₂ : ∀ x y z → z + (x + y) ≡ x + (y + z) + 0
+example₂ x y z = begin
+  z + (x + y) ≈⟨ +-comm z (x + y) ⟩
+  (x + y) + z ≈!⟨ solve +-0-monoid ⟩
```
[the remark this answers is not in the document]
Comment (oisdk) [reply README/MonoidSolver.agda]: It's the same problem as in #623 
Basically, when we call:

```agda
x + y ≈⟨ solve ⟩ y + z
```

The `solve` call needs to have the inferred type `x + y = y + z` *before* the reflection machinery gets going. With the standard combinator, type inference goes left to right, so it's instead inferred to have the type

```agda
x + y = ?
```

This different combinator flips the order of arguments.

I don't think there's a way to do it without a combinator like this one.

File: src/Algebra/Solver/Monoid/Reflection.agda [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+  go _ = nothing
+getArgs _ = nothing
+
+----------------------------------------------------------------------
+-- Getting monoid names
+----------------------------------------------------------------------
+
+-- We try to be flexible here, by matching two kinds of names.
+-- The first is the field accessor for the monoid record itself.
+-- However, users will likely want to use te solver with
```
Comment (MatthewDaggitt) [inline src/Algebra/Solver/Monoid/Reflection.agda]:
```suggestion
-- However, users will likely want to use the solver with
```

File: README/MonoidSolver.agda (hunk 4 above)
[the remark this answers is not in the document]
Comment (MatthewDaggitt) [reply README/MonoidSolver.agda]: Okay, thanks. Could you add that explanation as a comment above the new combinator? You should also add a similar combinator for `≡` I guess, as well as similar combinators in `Reasoning.Base.Double` and `Reasoning.Base.Triple`.

**Comment by MatthewDaggitt** [comment]
Also the comments at the top of the file explaining what's going on is fantastic. Thanks so much, I understand it much better now.

File: README/MonoidSolver.agda (hunk 4 above)
[the remark this answers is not in the document]
Comment (oisdk) [reply README/MonoidSolver.agda]: Or should I split it into a new pull request? There was a lot of discussion on #623 and this seems similarly sized if there's going to be a full duplication of the combinators into the inference-friendly form.

**Comment by JacquesCarette** [comment]
This is very cool. @alhassy and @ysharoda should also take a look at the main `Reflection` file and notice the boilerplate that's embedded in it.

**Comment by gallais** [comment]
Not blocked anymore, I believe.

**Comment by MatthewDaggitt** [comment]
Hi @oisdk, if you re-open this pull request against `experimental` instead of `master`, and add that comment to about the new reasoning combinator, I think this is ready to be merged in :+1:

**Comment by gallais** [comment]
I have changed the base branch on this PR. It does not seem to cause any issue.
I'll try to trigger a travis build to see whether it builds now.

**Comment by MatthewDaggitt** [comment]
Right made a bit of a hash of the commit history but this has finally been merged in! Thank you very much for this contribution @oisdk, sorry that it took so long to merge in. I have high hopes of seeing a similar pull request with the rest of your awesome ring solver soon :+1:

**Pull request merged**